### ISSUE TITLE
Memory Leak from FabricClient

### ISSUE THREAD

**Issue by danny156** [microsoft/service-fabric#330]
We have a SF stateless service which every 1 minute gets all applications in the cluster (Around ~700), iterate over their services in parallel tasks, and assert their health status.
The health status is logged to some provider.

However, every iteration the memory keeps increasing. After 1 day, it can reach to over 6GB. Inspecting the heap, we see this:
![Memory-Leak](https://user-images.githubusercontent.com/6977510/57783711-453c5800-7737-11e9-96f2-97485a156833.png)

Can you explain why CancellationToken exists so many times?
The service itself doesn't create any token, it just uses the one RunAsync gets.

We did a test where we passed CancellationToken.None to the entire flow (instead of the one received in the RunAsync), and memory seemed fine.

**Comment by GitTorre** [comment]
Thanks for reporting this issue.

How many instances of FabricClient are you using in your app health query code? Can you provide more information about your service impl?

**Comment by danny156** [comment]
Only one - class member.
This is generally how it looks (removed some code so it be clearer):


            var tasksList = new List<Task>();
            IEnumerbale<Application> applications = await GetClusterApplications();
            foreach (var application in applications)
            {

                cancellationToken.ThrowIfCancellationRequested();
                var servicesList = await GetApplicationServices(application);
                var task = Task.Run(async () =>
                {
                    foreach (var service in servicesList)
                    {
                        await _client.TestManager.ValidateServiceAsync(service, TimeSpan.FromSeconds(10), cancellationToken);
                    }
                }, cancellationToken);
                tasksList.Add(task);


            }

            await Task.WhenAll(tasksList.ToArray());`

**Comment by GitTorre** [comment]
Thanks for the snippet.  

Is there a reason you want to create a separate task for each call to ValidateServiceAsync?  If you have 700+ applications with n number of services, then you are using a lot of Task objects, which do carry some amount of cost (and, since the recommended practice is to not call Dispose on a Task (even though Task is IDisposable...), you aren't disposing the objects (not sure why this is the case, but like all things, you will want to make sure the recommendation actually applies to your specific usage scenario. I believe Stephen Toub has a blog post on the subject.)). 

If you want to do this in (mostly) parallel then use Parallel.For or Parallel.ForEach (and use Task.Factory for long running async work: Task.Factory.StartNew(() => Parallel.ForEach(....) ) 

That said, you aren't looking for a code review :-)... But, what makes you think this is a Service Fabric memory abuse issue? Are you seeing Fabric.exe consuming the memory? Or is your service process that does this work "leaking"?

**Comment by danny156** [comment]
Parallel.For or Parallel.ForEach losses their effect when used in async operations.
See: https://stackoverflow.com/questions/15136542/parallel-foreach-with-asynchronous-lambda

I am thinking it is a service fabric issue since the heap shows a lot of 'CancellationToken' objects. which are definitely not created by my service.

And anyway, I tried disposing the tasks -> didn't help.
I would expect a memory rise when all the tasks are running, but when they are disposed, I would expect the memory to return to normal. But this is isn't the case.

**Comment by maburlik** [comment]
Here's a sample of how to have Parallel.ForEach work for async operations:

```
public static class ParallelExtensions
    {
        /// <summary>
        /// From https://blogs.msdn.microsoft.com/pfxteam/2012/03/05/implementing-a-simple-foreachasync-part-2/
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="source"></param>
        /// <param name="dop">Degree of parallelism</param>
        /// <param name="body"></param>
        /// <returns></returns>
        public static Task ForEachAsync<T>(this IEnumerable<T> source, int dop, Func<T, Task> body)
        {
            return Task.WhenAll(
                from partition in Partitioner.Create(source).GetPartitions(dop)
                select Task.Run(async delegate {
                    using (partition)
                        while (partition.MoveNext())
                            await body(partition.Current);
                }));
        }
    }
```

intramethod:

```
await filesToUpload.ForEachAsync(_maxDegreeOfParallelism, async (string filepath) =>
                    {
                        DateTime lastModifiedTime = File.GetLastWriteTimeUtc(filepath);

                        if (lastModifiedTime < minDateTime)
                        {
                            return;
                        }

                        bool recordExists = uploadRecords.FileRecords.ContainsKey(filepath);

                        if (!recordExists)
                        {
                            uploadRecords.FileRecords[filepath] = new FileRecord() { FilePath = filepath, LastModifiedTime = lastModifiedTime };
                            cacheFileNeedsUpdate = true;
                        }
                        else if (recordExists
                            && uploadRecords.FileRecords[filepath].LastModifiedTime != lastModifiedTime)
                        {
                            uploadRecords.FileRecords[filepath].LastModifiedTime = lastModifiedTime;
                            cacheFileNeedsUpdate = true;
                        }
                        else
                        {
                            return;
                        }

                        string uploadPath = Path.Combine(containerUploadPath, Path.GetFileName(filepath));

                        long fileSize = new FileInfo(filepath).Length / 1024L;

                        TraceHelper.TraceDebug("Uploading {0} to {1}. Size: {2} KB. Modified: {3}", filepath, uploadPath, fileSize, lastModifiedTime);

                        CloudBlockBlob destinationBlob = container.GetBlockBlobReference(uploadPath);

                        using (FileStream fs = File.OpenRead(filepath))
                        {
                            await destinationBlob.UploadFromStreamAsync(fs);

                            Interlocked.Add(ref _totalUploadedKB, fileSize);
                        }
                    }).ConfigureAwait(false);
```

**Comment by GitTorre** [comment]
> Parallel.For or Parallel.ForEach losses their effect when used in async operations.
> See: https://stackoverflow.com/questions/15136542/parallel-foreach-with-asynchronous-lambda
> 
> I am thinking it is a service fabric issue since the heap shows a lot of 'CancellationToken' objects. which are definitely not created by my service.
> 
> And anyway, I tried disposing the tasks -> didn't help.
> I would expect a memory rise when all the tasks are running, but when they are disposed, I would expect the memory to return to normal. But this is isn't the case.

I still don't see the connection to our code. Any chance you could supply a working, small, sample that we could use to investigate? I realize this is probably not a simple undertaking, but without more information it's hard for us to verify your assumption.

**Comment by GitTorre** [comment]
OK. We have done some more internal analysis based on the information you have provided. It could be related to FabricClient, but we do need to spend more time exploring and ensuring we have the right info to share back with you. Give us some time to get back.

**Comment by danny156** [comment]
Sure, let me know if any extra info is needed.
@maburlik - Thanks, will give it a try.

**Comment by GitTorre** [comment]
@danny156 Can you try Disposing the FabricClient instance? Do you need to keep it around for the lifetime of the class instance? I think it would be interesting to see what happens if you were to Dispose the instance after the health check tasks complete and you move on to some other work. FabricClient can be used on-demand when you run your checks, dispose it when done, create an instance when you need it. It could still be a member of a class. The point of this exercise to see what happens with memory use (over time, the GC will collect when it does...) when you dispose the insance of Fabclient after each health check task blast. Make sense?

**Comment by MattHartz** [comment]
Hi, has this question been resolved? I have been experiencing issues with "TaskCancelledException" and I have reason to believe it is because our services are growing in memory ultimately resulting cancelled tasks firing.

I await tasks as they are created, is there more that needs to be done?
